ts/db.go

```diff
@@ -34,29 +38,106 @@ func NewDB(kv *client.KV) *DB {
 	}
 }
 
-// storeData attempts to store the supplied time series data on the server.
-// Data will be sampled at the supplied resolution.
-func (db *DB) storeData(r Resolution, data proto.TimeSeriesData) error {
-	internalData, err := data.ToInternal(r.KeyDuration(), r.SampleDuration())
-	if err != nil {
-		return err
+// PollSource begins a Goroutine which periodically queries the supplied
+// DataSource for time series data, storing the returned data in the server.
+// Stored data will be sampled using the provided Resolution. The polling
+// process will continue until the provided util.Stopper is stopped.
+func (db *DB) PollSource(source DataSource, frequency time.Duration, r Resolution, stopper *util.Stopper) {
+	p := &poller{
+		db:        db,
+		source:    source,
+		frequency: frequency,
+		r:         r,
+		stopper:   stopper,
+	}
+	p.start()
+}
+
+// A DataSource can be queryied for a slice of time series data.
+type DataSource interface {
+	GetTimeSeriesData() []proto.TimeSeriesData
+}
+
+// poller maintains information for a polling process started by PollSource().
+type poller struct {
+	db        *DB
+	source    DataSource
+	frequency time.Duration
+	r         Resolution
+	stopper   *util.Stopper
+}
+
+// start begins the goroutine for this poller, which will periodically request
+// time series data from the DataSource and store it.
+func (p *poller) start() {
+	p.stopper.RunWorker(func() {
+		ticker := time.NewTicker(p.frequency)
+		for {
+			select {
+			case <-ticker.C:
+				p.poll()
+			case <-p.stopper.ShouldStop():
+				return
+			}
+		}
+	})
+}
+
+// poll retrieves data from the underlying DataSource a single time, storing any
+// returned time series data on the server.
+func (p *poller) poll() {
+	if !p.stopper.StartTask() {
+		return
 	}
+	defer p.stopper.FinishTask()
 
-	for _, idata := range internalData {
-		key := MakeDataKey(data.Name, data.Source, r, idata.StartTimestampNanos)
-		value, err := idata.ToValue()
+	data := p.source.GetTimeSeriesData()
+	if len(data) == 0 {
+		return
+	}
+
+	if err := p.db.storeData(p.r, data...); err != nil {
```

Comment:
I see that the data is keyed by nanos, so it seems like you'd get away with having those things in goroutines. You could always limit the number of goroutines (logging errors if it's exceeded).